velox/functions/lib/Re2Functions.cpp

```diff
@@ -388,8 +388,143 @@ bool matchExactPattern(
     StringView input,
     const std::string& pattern,
     size_t length) {
-  return input.size() == pattern.size() &&
-      std::memcmp(input.data(), pattern.data(), length) == 0;
+  if (FOLLY_LIKELY(pattern.size() > 0)) {
+    return input.size() == pattern.size() &&
+        std::memcmp(input.data(), pattern.data(), length) == 0;
+  }
+
+  return input.size() == 0;
+}
+
+bool matchRelaxedFixedAscii(
+    StringView input,
+    const PatternMetadata& patternMetadata,
+    size_t start) {
+  // Compare the length first.
+  if (input.size() - start < patternMetadata.length()) {
+    return false;
+  }
+
+  for (const auto& subPattern : patternMetadata.subPatterns()) {
+    if (subPattern.kind == SubPatternKind::kLiteralString &&
+        std::memcmp(
+            input.data() + start + subPattern.start,
+            patternMetadata.fixedPattern().data() + subPattern.start,
+            subPattern.length) != 0) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool matchRelaxedFixedUnicode(
+    StringView input,
+    const PatternMetadata& patternMetadata,
+    size_t start) {
+  // Compare the length first.
+  if (input.size() - start < patternMetadata.length()) {
+    return false;
+  }
+
+  auto cursor = start;
+  for (const auto& subPattern : patternMetadata.subPatterns()) {
+    if (subPattern.kind == SubPatternKind::kSingleCharWildcard) {
+      // Match every single char wildcard.
+      for (auto i = 0; i < subPattern.length; i++) {
+        if (cursor >= input.size()) {
+          return false;
+        }
+
+        auto chrOffset = utf8proc_char_length(input.data() + cursor);
+        // Skip bad byte if we get utf length < 0.
```

Comment:
What is the behavior of RE2 in this case? Since this is an optimization, it would be nice to ensure that optimized and non-optimized behaviors match.

This code seems to repeat a few times. Perhaps, extract a helper method to skip N characters?

Reply:
> constructing an invalid byte stream manually?

That should work.